### ISSUE TITLE
feat: Implement `METADATA`

### ISSUE THREAD

**Issue by link2xt** [duesee/imap-codec#432]
There is currently no support for [METADATA](https://www.rfc-editor.org/rfc/rfc5464) extension.

Encoding SETMETADATA command and decoding METADATA responses returned for GETMETADATA commands does not seem to be possible currently.

**Comment by duesee** [comment]
[SORT/THREAD](https://github.com/duesee/imap-codec/pull/426) is next. Then [CONDSTORE/QRESYNC](https://github.com/duesee/imap-codec/pull/417). 

METADATA doesn't look too complicated. But the devil may be in the detail. Are you in a position to draft something up? :-)

If not, I can take a look in a few weeks.

**Comment by duesee** [comment]
# Capabilities

```abnf
capability        =/ "METADATA" / "METADATA-SERVER"
```

# Command

```abnf
command-auth      =/ setmetadata / getmetadata
```

## SETMETADATA

```abnf
; empty string for mailbox implies server annotation.
setmetadata       = "SETMETADATA" SP mailbox SP entry-values

entry-values      = "(" entry-value *(SP entry-value) ")"

entry-value       = entry SP value

; Slash-separated path to entry. MUST NOT contain "*" or "%".
; TODO(damian): astring allows them? Use `Wrapper(AString)`?
entry             = astring

value             = nstring / literal8
```

## GETMETADATA

```abnf
; empty string for mailbox implies server annotation.
getmetadata       = "GETMETADATA" [SP getmetadata-options] SP mailbox SP entries

getmetadata-options = "(" getmetadata-option *(SP getmetadata-option) ")"

; tagged-ext-label and tagged-ext-val are defined in [RFC4466].
; TODO(damian): See discussion in https://github.com/duesee/imap-codec/pull/417
getmetadata-option = tagged-ext-label [SP tagged-ext-val]

entries           = entry / "(" entry *(SP entry) ")"

; Used as a getmetadata-option
maxsize-opt       = "MAXSIZE" SP number

; Used as a getmetadata-option
scope-opt         =  "DEPTH" SP ("0" / "1" / "infinity")
```

# Response

```abnf
; empty string for mailbox implies server annotation.
metadata-resp     = "METADATA" SP mailbox SP (entry-values / entry-list)

entry-list        = entry *(SP entry)

response-payload  =/ metadata-resp

resp-text-code    =/ "METADATA" SP "LONGENTRIES" SP number / 
                     "METADATA" SP ("MAXSIZE" SP number / "TOOMANY" / "NOPRIVATE")
```

**Comment by duesee** [comment]
`imap-codec` doesn't do anything with connections. It only cares about parsing an input (and may return an error.) Connections will be handled by [`imap-flow`](https://github.com/duesee/imap-flow). Later, we'll have a higher-level library based on `imap-flow` that will be similar to `async-imap` (but based on a different operational model, libraries, etc.) Spoiler: IDLE will work better, it's cancellation-safe, you can use `select!`, etc.

Still, short answer: no. 

I'm not even sure how tolerating garbage in IMAP should work. IMAP doesn't really have "tokens" we could detect. The `go-imap` maintainer tried to tokenize the IMAP stream in v1 and gave up. (I hope I paraphrase him correctly.)

Thus, the best one could do is to detect lines and literals. But even then, this is asking for security issues.

Do you have examples of servers sending garbage? I don't mean like "unknown codes" or something that can be detected. But things similar to your `GARBAGE` response.

**Comment by duesee** [comment]
Thoughts:

> Most recent example is mail.163.com sends * STATUS "INBOX" () response which is explicitly prohibited by grammar which says status-att-list is non-empty.

That's unfortunate. The workaround is non-trivial because it would either require allowing an empty list (which would spread the incompatibility) or dropping the message (which would require additional logic).

Was this already reported to mail.163.com? Can't they just *stop* sending this?

> Another example was outlook server adding whitespace at the end of the line.

We should publicly report it to Microsoft. Workaround is easy and can be done w/o the risk to reproduce the fault (or changing semantics).

**Comment by duesee** [comment]
Hey @link2xt! Do you have an example how you use the metadata extension? I'm having a bit of a hard time figuring out how to present the entries to usere.

Currently, it's just an `astring`. But it could be a `PathBuf`, `Vec<Thing>`, `enum Entry { Private(...), ... ` etc.

The RFC is not totally clear to me. Any suggestions or experience?

**Comment by duesee** [comment]
Small update @link2xt: I talked with Arnt Gulbrandsen and asked him about ignoring/dropping bad messages. In practice, it may be safe to consume a line, see if it ends with `{...}`, if so, consume a literal, consume a line, see if it ends with `{...}`, if so, ... w/o doing IMAP parsing at all.

Funnily, I used this (seemingly hacky) approach already when implementing a server¹. But I changed it due to the danger of ambiguities. There are cases where `{...}` at the end of a line doesn't necessarily mean "literal", e.g., when `{` and `}` is used in a `Text`. But the insight is that these cases are rare, and, more importantly, can be reliably detected. It seems this is one of the many "undocumented truths" about IMAP and extensions stick to it.

TLDR;

Assuming that we can realiably detect literals means that we can split an IMAP stream into fragments w/o doing IMAP parsing at all. Thus, `imap-codec` could become more forgiving about garbage while still ensuring that garbage doesn't interfer with good messages.

¹ https://github.com/Email-Analysis-Toolkit/fake-mail-server
